tools/autograd/gen_variable_type.py

```diff
@@ -1918,20 +1942,30 @@ def emit_fw_derivatives() -> List[str]:
                     f"std::vector<{opt_res_grad_type}> {'_'.join(res)}_new_fw_grad_opts"
                     "(self.size(), c10::nullopt);"
                 )
-                inplace_foreach_forward_grad_formula = derivative.formula
-                for _foreach_arg, _ref_arg in inplace_foreacharg2refarg.items():
-                    # note(crcrpar): Massage only Scalar and ArrayRef<Scalar> here.
-                    if not (
-                        is_tensor_type(_foreach_arg.type)
-                        or is_tensor_list_type(_foreach_arg.type)
-                    ):
-                        pattern = _foreach_arg.name
-                        if isinstance(_foreach_arg.type, ListType):
-                            pattern += "[i]"
-                        inplace_foreach_forward_grad_formula = (
-                            inplace_foreach_forward_grad_formula.replace(
-                                _ref_arg.name, pattern
+                foreach_forward_grad_formula = derivative.formula
+                _foreach_arg: Union[Argument, DifferentiableInput]
+                if inplace:
+                    for _foreach_arg, _ref_arg in inplace_foreacharg2refarg.items():
+                        # note(crcrpar): Massage only Scalar and ArrayRef<Scalar> here.
+                        if not (
+                            is_tensor_type(_foreach_arg.type)
+                            or is_tensor_list_type(_foreach_arg.type)
+                        ):
+                            pattern = _foreach_arg.name
+                            if isinstance(_foreach_arg.type, ListType):
+                                pattern += "[i]"
+                            foreach_forward_grad_formula = (
+                                foreach_forward_grad_formula.replace(
+                                    _ref_arg.name, pattern
+                                )
```

Comment:
I'd still be curious why we need the "if inplace" here, and above in a couple places.

Currently, for the out of place case we are doing the replacements ahead of time, and for the inplace case, we are doing them in the gen_variable_type as we generate the code.

Whats preventing us from just not doing the replacement ahead of time in the out of place case? As that seems like it would allow us to avoid having extra if-else in gen_variable_type code.

Reply:
I think I try to re-use `NativeFunctionWithDifferentiabilityInfo` which is a frozen dataclass (as per https://github.com/pytorch/pytorch/blob/a200e3dfb8d0cc7699c3ccc1da64ef4ce7e0c0d1/torchgen/api/autograd.py#L253) for in-place foreach functions.
The motivation for this re-use is that in-place foreach creates grad_fn per Tensor while the out-place foreach creates grad_fn per function call. With this difference, I added the logic that modifies formula ahead of time and generates differentiability info for out-place, but not for in-place.

With that being said, it sounds worth thinking about unifying the two logics so that we have more modifications ahead of time (= in torchgen/api/autograd)

Reply:
> With that being said, it sounds worth thinking about unifying the two logics so that we have more modifications ahead of time (= in torchgen/api/autograd)

Yeah its possible that that results in simpler code, I'd be very happy to review any changes toward that!

I guess my question was that if we already have logic to modify the formula in-line for the in-place case, do we really need to do duplicate logic ahead of time and then also have to if/else to make sure we don't this work twice.

Reply:
one benefit of the reuse for in-place which I completely forgot is that `grad_fn`s of in-place foreach functions would have appropriate python bindings while those of out-place would not due to the lack of corresponding definitions in generated/Functions(.cpp|.h). I will try to figure out the way to push the modifications to torchgen though